src/corelibs/U2Core/src/util/MsaRowUtils.cpp

```diff
@@ -432,4 +432,80 @@ void MsaRowUtils::removeTrailingGapsFromModel(qint64 length, QVector<U2MsaGap> &
     }
 }
 
+QByteArray MsaRowUtils::getGappedSubsequence(const U2Region &region, const QByteArray &sequence, const QVector<U2MsaGap> &gaps) {
+    CHECK(region.length > 0, {});
+    CHECK(!sequence.isEmpty(), QByteArray(region.length, U2Msa::GAP_CHAR));
+    const char *coreSequence = sequence.constData();
+    int coreLength = sequence.length();
+    QByteArray result;
+    result.reserve(region.length);
+    int gapIndex = 0;  // Current gap index.
+    int corePos = 0;  // Current offset in core sequence.
+    int pos = 0;  // Current gapped position in the row.
+    int regionEnd = (int)region.endPos();  // Last (exclusive) global gapped position to include into the result.
+    if (region.startPos < 0) {
+        result.append(-region.startPos, U2Msa::GAP_CHAR);
+    }
+    // Iterate interleaved gap & core sequence regions.
+    while (gapIndex <= gaps.length() && pos < regionEnd && corePos < coreLength) {
+        const U2MsaGap *gap = gapIndex == gaps.length() ? nullptr : &gaps[gapIndex];
+        bool isValidGapModel = gapIndex == 0 || gap == nullptr || gap->startPos > gaps[gapIndex - 1].endPos();
```

Comment:
Wait, `gapIndex` equals to 0 on the first iteration, `gapIndex - 1` is `-1`, so `gaps[-1]` could cause a lot of problems here. Why isn't that happening?

Reply:
Oh, I see, `gapIndex == 0` is here. But don't you want to use `U2MsaGap::isValid()` too?

Reply:
How can I use U2MsaGap::isValid() here to compare 2 sibling gaps? Can you post an example, please?

Reply:
I'm not talking about comparing, just validate the `U2MsaGap` object

Reply:
What kind of error can be here with gap? Negative length should be handled in the gap constructor, but not in the 100+ places where gaps are used.